crates/air/src/components/memory.rs

```diff
@@ -0,0 +1,189 @@
+                qm31::PackedQM31,
+                SimdBackend,
+            },
+            BackendForChannel,
+        },
+        channel::{Channel, MerkleChannel},
+        fields::{m31::M31, qm31::SecureField, secure_column::SECURE_EXTENSION_DEGREE},
+        pcs::TreeVec,
+        poly::{
+            circle::{CanonicCoset, CircleEvaluation},
+            BitReversedOrder,
+        },
+    },
+};
+
+use crate::relations;
+
+#[derive(Copy, Clone, Default)]
+pub struct Claim {
+    pub log_size: u32,
+}
+
+#[derive(Copy, Clone)]
+pub struct InteractionClaim {
+    pub claimed_sum: SecureField,
+}
+
+pub struct LookupData {
+    pub memory: Vec<[PackedM31; 3]>,
+}
+
+impl Claim {
+    pub fn log_sizes(&self) -> TreeVec<Vec<u32>> {
+        let trace_log_sizes = vec![self.log_size; 3];
+        let interaction_log_sizes = vec![self.log_size; SECURE_EXTENSION_DEGREE];
+        TreeVec::new(vec![vec![], trace_log_sizes, interaction_log_sizes])
+    }
+
+    pub fn mix_into(&self, channel: &mut impl Channel) {
+        channel.mix_u64(self.log_size as u64);
+    }
+
+    #[allow(non_snake_case)]
+    pub fn write_trace<MC: MerkleChannel>(
+        &mut self,
+        lookup_data: &Vec<[PackedM31; 2]>,
+    ) -> (
+        [CircleEvaluation<SimdBackend, M31, BitReversedOrder>; 3],
+        LookupData,
+    )
+    where
+        SimdBackend: BackendForChannel<MC>,
+    {
+        let mut counts: HashMap<(u32, u32), u32> = HashMap::new();
+        for entry in lookup_data {
+            let arr0 = entry[0].to_array();
+            let arr1 = entry[1].to_array();
+            for (val0, val1) in arr0.iter().zip(arr1.iter()) {
+                let key = (val0.0, val1.0);
+                *counts.entry(key).or_insert(0) += 1;
+            }
+        }
+
+        let unique_pairs: Vec<_> = counts.into_iter().collect();
+        let mut addresses = Vec::new();
+        let mut values = Vec::new();
+        let mut mults = Vec::new();
+        let mut memory = Vec::new();
+
+        for chunk in unique_pairs.chunks(N_LANES) {
+            // If chunk is smaller than N_LANES, the remaining slots stay as M31(0)
+            let mut address = [M31(0); N_LANES];
+            let mut value = [M31(0); N_LANES];
+            let mut mult = [M31(0); N_LANES];
+
+            for (i, &((_address, _value), _mult)) in chunk.iter().enumerate() {
+                address[i] = M31(_address);
+                value[i] = M31(_value);
+                mult[i] = M31(_mult);
+            }
+
+            addresses.push(PackedM31::from_array(address));
+            values.push(PackedM31::from_array(value));
+            mults.push(PackedM31::from_array(mult));
+            memory.push([
+                addresses.last().unwrap().to_owned(),
+                values.last().unwrap().to_owned(),
+                mults.last().unwrap().to_owned(),
+            ]);
+        }
+
+        self.log_size = addresses.len().ilog2() + LOG_N_LANES;
```

Comment:
maybe here the log size is addresses.len().ilog2() - LOG_N_LANES
since we take chunks of the size of LOG_N_LANES?

Reply:
addresses is a vec of PackedM31, so each address counts for LOG_N_LANES entry, the size of the component is: size of address * (entry per packed m31), or log size = log size addresses + LOG_N_LANE;